core/Enclave/enclave_hsm.cpp

```diff
@@ -458,27 +540,111 @@ sgx_status_t enclave_rsa_encrypt(const uint8_t *cmk_blob,
+            }
+
+            if (ciphertext == NULL || ciphertext_len < RSA_OAEP_2048_CIPHER_LENGTH) {
+                printf("ecall rsa_encrypt ciphertext len is too small.\n");
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+            break;
+        case EH_RSA_3072:
+            sealed_rsa_len = sgx_calc_gcm_data_size(0, 3060);
+
+            if (plaintext_len > RSA_OAEP_3072_SHA_256_MAX_ENCRYPTION_SIZE) {
+                printf("ecall rsa_encrypt plain len is up to 318B.\n");
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+
+            if (ciphertext == NULL || ciphertext_len < RSA_OAEP_3072_CIPHER_LENGTH) {
+                printf("ecall rsa_encrypt ciphertext len is too small.\n");
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+            break;
+    }
+    
+    if (UINT32_MAX == sealed_rsa_len) {
+        printf("ecall rsa_encrypt failed to calculate sealed data size.\n");
+        return SGX_ERROR_UNEXPECTED;
+    }
+
+    if (cmk_blob == NULL || cmk_blob_size < sealed_rsa_len) {
+        printf("ecall rsa_encrypt cmk_blob_size is too small.\n");
         return SGX_ERROR_INVALID_PARAMETER;
     }
 
-    switch (keyspec)
-    {
-        case RSA_NO_PADDING:
-            RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, RSA_NO_PADDING);
-        case RSA_PKCS1_OAEP_PADDING:
-            RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, RSA_PKCS1_OAEP_PADDING);
+    if (plaintext == NULL || plaintext_len == 0) {
+        printf("ecall rsa_encrypt plaintext or len is wrong.\n");
+        return SGX_ERROR_INVALID_PARAMETER;
     }
 
-    return SGX_SUCCESS;
+    // encrypt
+    char* enc_key = NULL;
+    char* rsa_public_key = NULL;
+    BIO *bio = NULL;
+    RSA* rsa = NULL;
+    uint32_t public_key_size;
+    
+    do {
+        // load rsa public key
+        switch(keyspec)
+        {
+            case EH_RSA_2048:
+                public_key_size = 426;
+                break;
+            case EH_RSA_3072:
+                public_key_size = 601;
+                break;
+        }
+        enc_key = (char*)malloc(cmk_blob_size);
+        sgx_gcm_decrypt(&g_domain_key,
+                        (uint32_t*)&cmk_blob_size, (uint8_t *)enc_key,
+                        (sgx_aes_gcm_data_ex_t *)cmk_blob);
+
+        rsa_public_key = (char*)malloc(public_key_size);
+        memcpy(rsa_public_key, enc_key, public_key_size);
+
+        bio = BIO_new_mem_buf(rsa_public_key, -1);
+        if (bio == NULL) {
+            printf("failed to load rsa key pem\n");
+            break;
+        }
+
+        rsa = PEM_read_bio_RSAPublicKey(bio, NULL, NULL, NULL);
+        if (rsa == NULL) {
+            printf("failed to load rsa key\n");
+            break;
+        }
+
+        // make encryption
+        int res = RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, padding_mode);
+
+        if (!res)
```

Comment:
RSA_public_encrypt() returns the size of the encrypted data (i.e., RSA_size(rsa)).
On error, -1 is returned;

Reply:
Reruen check here does have a problem, I'll check it